content/post/2021-07-29-caching/index.Rmd

```diff
@@ -0,0 +1,223 @@
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+One principle of programming that's often encountered is "DRY", "Don't Repeat Yourself", that encourages e.g. the use of functions other duplicated (read: copy-pasted and slightly amended) code.
+You could also interpret it as don't let the machine repeat its calculations if useless.
+How about for a piece of code (function) with the same inputs, we only run it once per R session, and save the results for later?
+In this post, we shall go over ways to cache results in an R session, so that you don't need to burden machines.
+We will not cover [caching for R Markdown](https://bookdown.org/yihui/rmarkdown-cookbook/cache.html).
+
+## Caching: what is it and why use it?
+
+Caching means that if you call a function several times with the exact same input, the function is only actually run the first time.
+The result is stored in a cache of some sort (more practical details later!).
+Every other time the function is called with the same input, the result is retrieved from the cache.
+You will often think of caching as something valid in only one R session, but we'll see it can last longer via storage on disk.
+
+Now, *why* use caching?
+
+-   It might help save *time*.
+
+-   It might help save *other resources of users such as money*: e.g. if the function calls a web API whose pricing depends on the number of hits.
+    :sweat_smile:
+
+-   It might be *more polite*.
+    That's similar to the second item but from the perspective of e.g. a web API you keep hitting when you could have saved the result.
+    The [polite package](https://dmi3kno.github.io/polite/) for polite webscraping caches results.
+
+## Tools for caching in R
+
+Here's a roundup of ways to cache code in R.
+
+### The memoise package
+
+The [memoise package](https://memoise.r-lib.org/) by Jim Hester is easy to use.
+Say we want to cache a function that only sleeps.
+
+```{r memoise}
+.sleep <- function() {
+  Sys.sleep(3)
+  "Rested now!"
+}
+
+sleep <- memoise::memoise(.sleep)
+
+system.time(sleep())
+system.time(sleep())
+```
+
+The second call to `sleep()` is much quicker because, well, it does not call the `.sleep()` function so there's no sleep.
+
+The memoise package also lets you
+
+-   choose the duration of validity of the cache;
+
+-   cache on disk -- on that topic see the R-hub blog post on [persistent data and config for R packages](/2020/03/12/user-preferences/).
+
+If you use the memoise package in a package, do not forget to add
+
+``` r
+@importFrom memoise memoise
+```
+
+in one of your R scripts (thanks [Mark Padgham](https://mpadge.github.io/) for this tip!) otherwise you will get a R CMD Check NOTE.
+
+    Result: NOTE
+        Namespace in Imports field not imported from: ‘memoise’
+         All declared Imports should be used.
```

Comment:
added a sentence to clarify " This is because R will look for package usage in function bodies, whereas the call to memoise is at the top-level." (rephrased from a comment of Jim's in rOpenSci Slack)